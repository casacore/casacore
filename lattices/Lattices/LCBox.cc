//# LCBox.cc: Class to define a rectangular box of interest
//# Copyright (C) 1997,1998,1999,2001,2003
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <lattices/Lattices/LCBox.h>
#include <casa/BasicMath/Math.h>
#include <casa/Arrays/ArrayMath.h>
#include <tables/Tables/TableRecord.h>
#include <casa/Utilities/Assert.h>
#include <casa/Exceptions/Error.h>
#include <casa/iostream.h>
#include <casa/sstream.h>

namespace casa { //# NAMESPACE CASA - BEGIN

LCBox::LCBox()
{}

LCBox::LCBox (const IPosition& latticeShape)
: LCRegionFixed (latticeShape)
{
    // Set the box to the full lattice.
    setBoundingBox (Slicer (IPosition(latticeShape.nelements(), 0),
			    latticeShape));
    // Fill the blc and trc vectors.
    fillBlcTrc();
}

LCBox::LCBox (const Slicer& box, const IPosition& latticeShape)
: LCRegionFixed (latticeShape)
{
    // Make sure no stride is given.
    if (box.stride() != 1) {
	throw (AipsError ("LCBox::LCBox - "
			  "stride in given Slicer has to be 1"));
    }
    // When the slicer is fixed (i.e. blc and trc explicitly given),
    // it is possible that it partly exceeds the lattice boundaries.
    if (box.isFixed()) {
        setSlicerBox (box.start(), box.end());
    } else {
	setBoundingBox (box);
    }
    // Fill the blc and trc vectors.
    fillBlcTrc();
}

// Construct from the IPosition's defining the bottom-left and
// top-right corner of the box.
LCBox::LCBox (const IPosition& blc, const IPosition& trc,
	      const IPosition& latticeShape)
: LCRegionFixed (latticeShape)
{
    setSlicerBox (blc, trc);
    fillBlcTrc();
}

LCBox::LCBox (const Vector<Float>& blc, const Vector<Float>& trc,
	      const IPosition& latticeShape)
: LCRegionFixed (latticeShape),
  itsBlc        (blc.copy()),
  itsTrc        (trc.copy())
{
    uInt i;
    IPosition bl(blc.nelements());
    for (i=0; i<blc.nelements(); i++) {
	bl(i) = Int(blc(i) + 0.5);
    }
    IPosition tr(trc.nelements());
    for (i=0; i<trc.nelements(); i++) {
	tr(i) = Int(trc(i) + 0.5);
    }
    setSlicerBox (bl, tr);
}

LCBox::LCBox (const Vector<Double>& blc, const Vector<Double>& trc,
	      const IPosition& latticeShape)
: LCRegionFixed (latticeShape),
  itsBlc        (blc.nelements()),
  itsTrc        (trc.nelements())
{
    uInt i;
    IPosition bl(blc.nelements());
    for (i=0; i<blc.nelements(); i++) {
	itsBlc(i) = blc(i);
	bl(i) = Int(blc(i) + 0.5);
    }
    IPosition tr(trc.nelements());
    for (i=0; i<trc.nelements(); i++) {
	itsTrc(i) = trc(i);
	tr(i) = Int(trc(i) + 0.5);
    }
    setSlicerBox (bl, tr);
}

LCBox::LCBox (const LCBox& other)
: LCRegionFixed (other),
  itsBlc(other.itsBlc),
  itsTrc(other.itsTrc)
{}
 

LCBox::~LCBox()
{}

LCBox& LCBox::operator= (const LCBox& other)
{
    if (this != &other) {
	LCRegionFixed::operator= (other);
	itsBlc.resize(other.itsBlc.nelements());
	itsTrc.resize(other.itsTrc.nelements());
	itsBlc = other.itsBlc;    
	itsTrc = other.itsTrc;
    }
    return *this;
}


Bool LCBox::operator== (const LCRegion& other) const
{
    // Check if parent class matches.
    // If so, we can safely cast.
    if (! LCRegionFixed::operator== (other)) {
	return False;
    }
    const LCBox& that = (const LCBox&)other;
    // Compare private data.
    if (itsBlc.nelements() != that.itsBlc.nelements()
    ||  itsTrc.nelements() != that.itsTrc.nelements()) {
	return False;
    }
    for (uInt i=0; i<itsBlc.nelements(); i++) {
	if (!near (itsBlc(i), that.itsBlc(i))
	||  !near (itsTrc(i), that.itsTrc(i))) {
	    return False;
	}
    }
    return True;
}

LCRegion* LCBox::cloneRegion() const
{
    return new LCBox(*this);
}

LCRegion* LCBox::doTranslate (const Vector<Float>& translateVector,
			      const IPosition& newLatticeShape) const
{
    uInt ndim = latticeShape().nelements();
    Vector<Float> blc (itsBlc.copy());
    Vector<Float> trc (itsTrc.copy());
    for (uInt i=0; i<ndim; i++) {
        blc(i) += translateVector(i);
        trc(i) += translateVector(i);
    }
    return new LCBox (blc, trc, newLatticeShape);
}

String LCBox::className()
{
    return "LCBox";
}

String LCBox::type() const
{
   return className();
}

TableRecord LCBox::toRecord (const String&) const
{
    TableRecord rec;
    defineRecordFields (rec, className());
    // Write 1-relative.
    rec.define ("oneRel", True);
    rec.define ("blc", itsBlc + Float(1));
    rec.define ("trc", itsTrc + Float(1));
    rec.define ("shape", latticeShape().asVector());
    return rec;
}

LCBox* LCBox::fromRecord (const TableRecord& rec, const String&)
{
    // If 1-relative, subtract 1 from blc and trc.
    Bool oneRel = rec.asBool ("oneRel");
    Float off = (oneRel ? 1:0);
    Array<Float> blc (rec.toArrayFloat ("blc"));
    Array<Float> trc (rec.toArrayFloat ("trc"));
    return new LCBox (blc-off, trc-off,
		      Vector<Int>(rec.toArrayInt ("shape")));
}

void LCBox::setSlicerBox (const IPosition& blc, const IPosition& trc)
{
    const IPosition& shape = latticeShape();
    uInt ndim = shape.nelements();
    if (blc.nelements() != ndim  ||  trc.nelements() != ndim) {
	throw (AipsError ("LCBox::LCBox - "
			  "length of blc and trc vectors have to match "
			  "dimensionality of lattice"));
    }
    IPosition bl(blc);
    IPosition tr(trc);
    for (uInt i=0; i<ndim; i++) {
        if (bl(i) < 0) {
	    bl(i) = 0;
	}
	if (tr(i) >= shape(i)) {
	    tr(i) = shape(i) - 1;
	}
	if (bl(i) > tr(i)) {
	    ostringstream bstr, tstr;
	    bstr << bl;
	    tstr << tr;
	    throw (AipsError ("LCBox::LCBox - "
			      "blc " + String(bstr) + " must be <= trc "
			      + String(tstr)));
	}
    }
    setBoundingBox (Slicer(bl, tr, Slicer::endIsLast));
}

void LCBox::fillBlcTrc()
{
    const Slicer& sl = boundingBox();
    uInt nd = sl.ndim();
    itsBlc.resize (nd);
    itsTrc.resize (nd);
    for (uInt i=0; i<nd; i++) {
	itsBlc(i) = sl.start()(i);
	itsTrc(i) = sl.end()(i);
    }
}

Bool LCBox::verify (IPosition& blc, IPosition& trc,
                    IPosition& inc, const IPosition& shape)
{
   IPosition inBlc(blc);
   IPosition inTrc(trc);
   IPosition inInc(inc);
   const Int nDim = shape.nelements();

// Check blc

   const Int blcDim = blc.nelements();
   blc.resize(nDim,True);
   if (blcDim == 0) {
      blc = 0;
   } else {
      for (Int i=0; i<nDim; i++) {
         if (i > blcDim-1) {
            blc(i) = 0;
         } else {
            if (blc(i) < 0 || blc(i) > shape(i)-1) blc(i) = 0;
         }
      }
   }

// Check trc

   const Int trcDim = trc.nelements();
   trc.resize(nDim,True);
   if (trcDim == 0) {
      trc = shape- 1;
   } else {
      for (Int i=0; i<nDim; i++) {
         if (i > trcDim-1) {
            trc(i) = shape(i) - 1;
         } else {
            if (trc(i) < 0 || trc(i) > shape(i)-1) {
               trc(i) = shape(i) - 1;
            }
         }
      }
   }

// Check increment

   const Int incDim = inc.nelements();
   inc.resize(nDim,True);
   if (incDim == 0) {
      inc = 1;
   } else {
      for (Int i=0; i<nDim; i++) {
         if (i > incDim-1) {
            inc(i) = 1;
         } else {
            if (inc(i) < 1 || inc(i) > trc(i)-blc(i)+1) inc(i) = 1;
         }
      }     
   }     

// Check blc<trc 

   for (Int i=0; i<nDim; i++) {
      if (blc (i) > trc(i)) {
         blc(i) = 0;
         trc(i) = shape(i) - 1;
      }
   }
//
   Bool changed = (blc.nelements()!=inBlc.nelements() ||
                         trc.nelements()!=inTrc.nelements() ||
                         inc.nelements()!=inInc.nelements());
   if (!changed) changed = (blc!=inBlc || trc!=inTrc || inc!=inInc);
//
   return changed;
}
  

             


} //# NAMESPACE CASA - END

