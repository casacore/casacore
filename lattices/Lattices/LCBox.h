//# LCBox.h: Class to define a rectangular box of interest
//# Copyright (C) 1997,1998,1999
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef LATTICES_LCBOX_H
#define LATTICES_LCBOX_H

//# Includes
#include <casacore/casa/aips.h>
#include <casacore/lattices/Lattices/LCRegionFixed.h>
#include <casacore/casa/Arrays/Slicer.h>
#include <casacore/casa/Arrays/Vector.h>


namespace casacore { //# NAMESPACE CASACORE - BEGIN

// <summary>
// Class to define a rectangular box of interest.
// </summary>

// <use visibility=export>

// <reviewed reviewer="" date="" tests="">
// </reviewed>

// <prerequisite>
//   <li> <linkto class=LCRegion>LCRegion</linkto>
// </prerequisite>

// <synopsis> 
// The LCBox class is a specialization of class
// <linkto class=LCRegion>LCRegion</linkto>.
// It makes it possible to define a rectangular region of interest.
// </synopsis> 

// <example>
// <srcblock>
// </srcblock>
// </example>

// <todo asof="1997/11/11">
// </todo>

class LCBox: public LCRegionFixed
{
public:
    LCBox();

    // Construct a box for the full lattice shape.
    explicit LCBox (const IPosition& latticeShape);

    // Construct from the Slicer defining the box.
    // The slicer may not contain a stride.
    LCBox (const Slicer& box, const IPosition& latticeShape);

    // Construct from the IPosition's defining the bottom-left and
    // top-right corner of the box.
    LCBox (const IPosition& blc, const IPosition& trc,
	   const IPosition& latticeShape);

    // Construct from the Vector's defining the bottom-left and
    // top-right corner of the box.
    // <group>
    LCBox (const Vector<Float>& blc, const Vector<Float>& trc,
	   const IPosition& latticeShape);
    LCBox (const Vector<Double>& blc, const Vector<Double>& trc,
	   const IPosition& latticeShape);
    // </group>

    // Copy constructor (reference semantics).
    LCBox (const LCBox& other);

    virtual ~LCBox();

    // Assignment (copy semantics).
    LCBox& operator= (const LCBox& other);

    // Comparison.  Mask not checked. Use function 
    // LRegionSingle::maskEqual  to do this
    virtual Bool operator== (const LCRegion& other) const;

    // Make a copy of the derived object.
    virtual LCRegion* cloneRegion() const;

    // Get the class name (to store in the record).
    static String className();

    // Get the region type.  Returns className()
    virtual String type() const;

    // Convert the (derived) object to a record.
    virtual TableRecord toRecord (const String& tableName) const;

    // Convert correct object from a record.
    static LCBox* fromRecord (const TableRecord&,
			      const String& tablename);

    // Get the box blc
    Vector<Float> blc() const;

    // Get the box trc
    Vector<Float> trc() const;

// Verify a box specification.  Illegal (inlcuding blc > trc) or
// unspecified values are  given 0 (blc) shape (trc) or
// unity (inc).  Returns <src>True</src> if any of the blc/trc/inc 
// are changed from their input values, else returns <src>False</src>
   static Bool verify (IPosition& blc, IPosition& trc,
                       IPosition& inc, const IPosition& shape);


protected:
    // Construct another LCBox (for e.g. another lattice) by moving
    // this one. It recalculates the bounding box.
    // A positive translation value indicates "to right".
    virtual LCRegion* doTranslate (const Vector<Float>& translateVector,
				   const IPosition& newLatticeShape) const;

private:
    // Make a box from the blc,trc such that it does not exceed the
    // lattice boundaries.
    void setSlicerBox (const IPosition& blc, const IPosition& trc);

    // Fill the blc and trc vector from IPositions.
    void fillBlcTrc();


    //# Variables
    Vector<Float> itsBlc;
    Vector<Float> itsTrc;
};


inline Vector<Float> LCBox::blc() const
{
    return itsBlc;
}
inline Vector<Float> LCBox::trc() const
{
    return itsTrc;
}



} //# NAMESPACE CASACORE - END

#endif
