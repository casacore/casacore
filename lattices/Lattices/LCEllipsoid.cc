//# LCEllipsoid.cc: Define an N-dimensional ellipsoidal region of interest
//# Copyright (C) 1997,1998,1999,2001,2003
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$


#include <lattices/Lattices/LCEllipsoid.h>
#include <tables/Tables/TableRecord.h>
#include <casa/Arrays/ArrayMath.h>
#include <casa/Arrays/ArrayIO.h>
#include <casa/Exceptions/Error.h>
#include <casa/Utilities/Assert.h>
#include <casa/iostream.h>
#include <casa/sstream.h>

namespace casa { //# NAMESPACE CASA - BEGIN

LCEllipsoid::LCEllipsoid()
{}

LCEllipsoid::LCEllipsoid (const IPosition& center, Float radius,
			  const IPosition& latticeShape)
: LCRegionFixed (latticeShape),
  itsRadii      (latticeShape.nelements())
{
    fillCenter (center);
    itsRadii = radius;
    setBoundingBox (makeBox (itsCenter, itsRadii, latticeShape));
    defineMask();
}

LCEllipsoid::LCEllipsoid (const Vector<Float>& center, Float radius,
			  const IPosition& latticeShape)
: LCRegionFixed (latticeShape),
  itsCenter     (center.copy()),
  itsRadii      (latticeShape.nelements())
{
    itsRadii = radius;
    setBoundingBox (makeBox (itsCenter, itsRadii, latticeShape));
    defineMask();
}

LCEllipsoid::LCEllipsoid (const Vector<Double>& center, Double radius,
			  const IPosition& latticeShape)
: LCRegionFixed (latticeShape),
  itsCenter     (center.nelements()),
  itsRadii      (center.nelements())
{
    for (uInt i=0; i<center.nelements(); i++) {
	itsCenter(i) = center(i);
	itsRadii(i) = radius;
    }
    setBoundingBox (makeBox (itsCenter, itsRadii, latticeShape));
    defineMask();
}

LCEllipsoid::LCEllipsoid (const Vector<Float>& center,
			  const Vector<Float>& radii,
			  const IPosition& latticeShape)
: LCRegionFixed (latticeShape),
  itsCenter     (center.copy()),
  itsRadii      (radii.copy())
{
    setBoundingBox (makeBox (itsCenter, itsRadii, latticeShape));
    defineMask();
}

LCEllipsoid::LCEllipsoid (const Vector<Double>& center,
			  const Vector<Double>& radii,
			  const IPosition& latticeShape)
: LCRegionFixed (latticeShape),
  itsCenter     (center.nelements()),
  itsRadii      (radii.nelements())
{
    for (uInt i=0; i<center.nelements(); i++) {
	itsCenter(i) = center(i);
	if (i < radii.nelements()) {
	    itsRadii(i) = radii(i);
	}
    }
    setBoundingBox (makeBox (itsCenter, itsRadii, latticeShape));
    defineMask();
}

LCEllipsoid::LCEllipsoid (const LCEllipsoid& other)
: LCRegionFixed (other),
  itsCenter     (other.itsCenter),
  itsRadii      (other.itsRadii)
{}

LCEllipsoid::~LCEllipsoid()
{}

LCEllipsoid& LCEllipsoid::operator= (const LCEllipsoid& other)
{
    if (this != &other) {
	LCRegionFixed::operator= (other);
	itsCenter.resize (other.itsCenter.nelements());
	itsRadii.resize  (other.itsCenter.nelements());
	itsCenter = other.itsCenter;
	itsRadii  = other.itsRadii;
    }
    return *this;
}

Bool LCEllipsoid::operator== (const LCRegion& other) const
{
    // Check if parent class matches.
    // If so, we can safely cast.
    if (! LCRegionFixed::operator== (other)) {
	return False;
    }
    const LCEllipsoid& that = (const LCEllipsoid&)other;
    // Compare private data.
    if (itsCenter.nelements() != that.itsCenter.nelements()
    ||  itsRadii.nelements()  != that.itsRadii.nelements()) {
	return False;
    }
    for (uInt i=0; i<itsCenter.nelements(); i++) {
	if (!near (itsCenter(i), that.itsCenter(i))
	||  !near (itsRadii(i),  that.itsRadii(i))) {
	    return False;
	}
    }
    return True;
}


LCRegion* LCEllipsoid::cloneRegion() const
{
    return new LCEllipsoid(*this);
}

LCRegion* LCEllipsoid::doTranslate (const Vector<Float>& translateVector,
				    const IPosition& newLatticeShape) const
{
    uInt ndim = latticeShape().nelements();
    Vector<Float> center;
    center = itsCenter;
    for (uInt i=0; i<ndim; i++) {
        center(i) += translateVector(i);
    }
    return new LCEllipsoid (center, itsRadii, newLatticeShape);
}

String LCEllipsoid::className()
{
    return "LCEllipsoid";
}

String LCEllipsoid::type() const
{
   return className();
}

TableRecord LCEllipsoid::toRecord (const String&) const
{
    TableRecord rec;
    defineRecordFields (rec, className());
    // Write 1-relative.
    rec.define ("oneRel", True);
    rec.define ("center", itsCenter + Float(1));
    rec.define ("radii", itsRadii);
    rec.define ("shape", latticeShape().asVector());
    return rec;
}

LCEllipsoid* LCEllipsoid::fromRecord (const TableRecord& rec,
				      const String&)
{
    // If 1-relative, subtract 1 from center.
    Bool oneRel = rec.asBool ("oneRel");
    Float off = (oneRel ? 1:0);
    Array<Float> center (rec.toArrayFloat ("center"));
    return new LCEllipsoid (center-off,
			    Vector<Float>(rec.toArrayFloat ("radii")),
			    Vector<Int>(rec.toArrayInt ("shape")));
}

void LCEllipsoid::fillCenter (const IPosition& center)
{
    itsCenter.resize (center.nelements());
    for (uInt i=0; i<center.nelements(); i++) {
	itsCenter(i) = center(i);
    }
}

Slicer LCEllipsoid::makeBox (const Vector<Float>& center,
			     const Vector<Float>& radii,
			     const IPosition& latticeShape)
{
    uInt nrdim = center.nelements();
    // First make sure dimensionalities conform.
    if (latticeShape.nelements() != nrdim  ||  radii.nelements() != nrdim) {
	throw (AipsError ("LCEllipsoid::LCEllipsoid - "
			  "dimensionality of center,radii,lattice mismatch"));
    }
    // Determine blc and trc.
    IPosition blc(nrdim);
    IPosition trc(nrdim);
    for (uInt i=0; i<nrdim; i++) {
	if (center(i) > latticeShape(i)-1  ||  center(i) < 0) {
	    ostringstream cstr, lstr;
	    cstr << center;
	    lstr << latticeShape;
	    throw (AipsError ("LCEllipsoid::LCEllipsoid - "
			      "invalid center " + String(cstr) +
			      " (outside lattice " + String(lstr) + ")"));
	}
	blc(i) = Int(center(i) - radii(i) + 0.5);
	if (blc(i) < 0) {
	    blc(i) = 0;
	}
	trc(i) = Int(center(i) + radii(i) + 0.5);
	if (trc(i) >= latticeShape(i)) {
	    trc(i) = latticeShape(i) - 1;
	}
	if (blc(i) > trc(i)) {
	    ostringstream rstr;
	    rstr << radii;
	    throw (AipsError ("LCEllipsoid::LCEllipsoid - "
			      "ellipsoid is empty (radii " + String(rstr) +
			      " too small)"));
	}
    }
    // In principle it is possible that e.g. the top of the ellipsoid falls
    // between 2 grid points, so we could leave out that line of pixels.
    return Slicer(blc, trc, Slicer::endIsLast);
}

void LCEllipsoid::defineMask()
{
    uInt i;
    // Create the mask with the shape of the bounding box.
    // Set the mask initially to False.
    const IPosition& length = boundingBox().length();
    uInt nrdim = length.nelements();
    Array<Bool> mask(length);
    mask = False;
    // Get access to the mask storage.
    Bool deleteIt;
    Bool* maskData = mask.getStorage (deleteIt);
    // Initialize some variables for the loop below.
    Float center0 = itsCenter(0) - boundingBox().start()(0);
    Float radsq0 = itsRadii(0) * itsRadii(0);
    Int np = length(0);
    IPosition pos (nrdim, 0);
    Vector<Float> center (nrdim);
    Vector<Float> radsq (nrdim);
    Vector<Float> dist (nrdim, 0.0);
    Float distsq = 0;
    for (i=1; i<nrdim; i++) {
	center(i) = itsCenter(i) - boundingBox().start()(i);
	Float d = max (float(0), center(i)-0.5f) / itsRadii(i);
	dist(i) = d * d;
	distsq += dist(i);
    }
    // Loop through all pixels in the ellipsoid.
    // The outer loop iterates over all lines.
    // The inner loop sets the pixel mask for a line by calculating
    // the start and end of the ellipsoid for that line.
    // The variable distsq contains the 'distance' of the line to the center.
    for (;;) {
	// Ignore the line when the distance exceeds the radius.
	Float d = 1 - distsq;
	if (d >= 0) {
	    d = sqrt(d * radsq0);
	    Int st = Int(center0 - d + 0.5);
	    if (st < 0) {
		st = 0;
	    }
	    Int end = Int(center0 + d + 0.5);
	    if (end >= np) {
		end = np-1;
	    }
	    for (Int j=st; j<=end; j++) {
		maskData[j] = True;
	    }
	}
	// Go to the next line and update the line distance.
	maskData += np;
	for (i=1; i<nrdim; i++) {
	    distsq -= dist(i);
	    if (++pos(i) < length(i)) {
		Float d = abs(center(i) - pos(i));
		d = max(float(0), d-0.5f) / itsRadii(i);
		dist(i) = d*d;
		distsq += dist(i);
		break;
	    }
	    // This dimension is done. Reset it and continue with the next.
	    pos(i) = 0;
	    Float d = max (float(0), center(i)-0.5f) / itsRadii(i);
	    dist(i) = d*d;
	    distsq += dist(i);
        }
	// End the iteration when all dimensions are done.
	if (i == nrdim) {
	    break;
	}
	DebugAssert (maskData == &mask(pos), AipsError);
    }
    mask.putStorage (maskData, deleteIt);
    setMask (mask);
}

} //# NAMESPACE CASA - END

