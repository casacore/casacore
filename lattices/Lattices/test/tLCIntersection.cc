//# tLCIntersection.cc:  mechanical test of the LCIntersection class
//# Copyright (C) 1998,1999,2000,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <lattices/Lattices/LCIntersection.h>
#include <lattices/Lattices/LCBox.h>
#include <lattices/Lattices/LCEllipsoid.h>
#include <casa/Arrays/ArrayIO.h>
#include <casa/Arrays/ArrayLogical.h>
#include <casa/Arrays/IPosition.h>
#include <tables/Tables/TableRecord.h>
#include <casa/Utilities/Assert.h>
#include <casa/Exceptions/Error.h>
#include <casa/iostream.h>


#include <casa/namespace.h>
void doIt (const IPosition& latticeShape,
	   const IPosition& start,
	   const IPosition& end,
	   const IPosition& center,
	   Float radius)
{
    uInt ndim = latticeShape.nelements();
    LCBox box (start, end, latticeShape);
    LCEllipsoid cir (center, radius, latticeShape);
    LCIntersection inters (box, cir);
    AlwaysAssertExit (inters.hasMask());
    AlwaysAssertExit (! inters.isWritable());
    cout << inters.hasMask() << ' ' << endl;
    cout << inters.boundingBox().start() << inters.boundingBox().end()
	 << inters.boundingBox().length() << inters.latticeShape() << endl;
    Array<Bool> mask;
    inters.getSlice (mask, IPosition(ndim,0), inters.boundingBox().length(),
		     IPosition(ndim,1));
    cout << mask << endl;

    LCIntersection inters1 (False, &box); 
    AlwaysAssertExit (! inters1.hasMask());
    AlwaysAssertExit (! inters1.isWritable());
    Array<Bool> mask1;
    inters1.getSlice (mask1, IPosition(ndim,0), inters1.boundingBox().length(),
		     IPosition(ndim,1));
    AlwaysAssertExit (allEQ(mask1, True));
   
    {
	// Test cloning.
        LCRegion* interscop = inters.cloneRegion();
	AlwaysAssertExit (inters.hasMask() == interscop->hasMask());
	AlwaysAssertExit (inters.boundingBox().start() ==
			  interscop->boundingBox().start());
	AlwaysAssertExit (inters.boundingBox().end() ==
			  interscop->boundingBox().end());
	AlwaysAssertExit (inters.boundingBox().stride() ==
			  interscop->boundingBox().stride());
	AlwaysAssertExit (inters.boundingBox().length() ==
			  interscop->boundingBox().length());
	Array<Bool> arr;
	interscop->getSlice (arr, IPosition(ndim,0),
			     inters.boundingBox().length(),
			     IPosition(ndim,1));
	AlwaysAssertExit (allEQ (arr, mask));
	delete interscop;
    }
    {
	// Test persistency.
	LCRegion* interscop = (LCRegion::fromRecord (inters.toRecord(""), ""));
	AlwaysAssertExit (inters.hasMask() == interscop->hasMask());
	AlwaysAssertExit (inters.boundingBox().start() ==
			  interscop->boundingBox().start());
	AlwaysAssertExit (inters.boundingBox().end() ==
			  interscop->boundingBox().end());
	AlwaysAssertExit (inters.boundingBox().stride() ==
			  interscop->boundingBox().stride());
	AlwaysAssertExit (inters.boundingBox().length() ==
			  interscop->boundingBox().length());
	Array<Bool> arr;
	interscop->getSlice (arr, IPosition(ndim,0),
			     inters.boundingBox().length(),
			     IPosition(ndim,1));
	AlwaysAssertExit (allEQ (arr, mask));
	delete interscop;
    }
    {
    // Test ordered equality.
       LCIntersection inters1(box, cir);
       LCIntersection inters2(inters1);
       AlwaysAssertExit (inters1 == inters2);
    }
    {
    // Test unordered equality.
       LCIntersection inters1(box, cir);
       LCIntersection inters2(cir, box);
       AlwaysAssertExit (inters1 == inters2);
    }   
}


int main()
{
    try {
	doIt (IPosition (2,11,20),
	      IPosition (2,3,4), IPosition (2,7,8),
	      IPosition (2,5,10), 5.);
	try {
	    doIt (IPosition (2,10,20),
		  IPosition (2,3,4), IPosition (2,7,8),
		  IPosition (2,4,16), 5.);
	} catch (AipsError x) {
	    cout << x.getMesg() << endl;
	} 
    } catch (AipsError x) {
	cout << "Caught exception: " << x.getMesg() << endl;
	return 1;
    } 
    cout << "OK" << endl;
    return 0;
}
