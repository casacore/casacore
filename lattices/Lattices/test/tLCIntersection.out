1 
[3, 5][7, 8][5, 4][11, 20]
Axis Lengths: [5, 4]  (NB: Matrix in Row/Column order)
[1, 1, 1, 1
 1, 1, 1, 1
 1, 1, 1, 1
 1, 1, 1, 1
 1, 1, 1, 1]

LCIntersection::LCIntersection - regions do not overlap
OK
